vpr/src/place/net_cost_handler.cpp

```diff
@@ -0,0 +1,2277 @@
+                            t_2D_bb& bb_coord_new);
+
+/**
+ * @brief Calculate the 3D BB of a large net from scratch and update coord, edge, and num_sink_pin_layer data structures.
+ * @param net_id
+ * @param coords
+ * @param num_on_edges
+ * @param num_sink_pin_layer
+ */
+static void get_bb_from_scratch(ClusterNetId net_id,
+                                t_bb& coords,
+                                t_bb& num_on_edges,
+                                vtr::NdMatrixProxy<int, 1> num_sink_pin_layer);
+
+/**
+ * @brief Calculate the per-layer BB of a large net from scratch and update coord, edge, and num_sink_pin_layer data structures.
+ * @param net_id
+ * @param coords
+ * @param num_on_edges
+ * @param num_sink_pin_layer
+ */
+static void get_layer_bb_from_scratch(ClusterNetId net_id,
+                                      std::vector<t_2D_bb>& num_on_edges,
+                                      std::vector<t_2D_bb>& coords,
+                                      vtr::NdMatrixProxy<int, 1> layer_pin_sink_count);
+
+/**
+ * @brief Given the 3D BB, calculate the wire-length cost of the net
+ * @param net_id
+ * @param bb
+ * @return
+ */
+static double get_net_cost(ClusterNetId net_id, const t_bb& bb);
+
+/**
+ * @brief Given the per-layer BB, calculate the wire-length cost of the net on each layer
+ * and return the sum of the costs
+ * @param net_id
+ * @param bb
+ * @return
+ */
+static double get_net_layer_cost(ClusterNetId /* net_id */,
+                                 const std::vector<t_2D_bb>& bb,
+                                 const vtr::NdMatrixProxy<int, 1> layer_pin_sink_count);
+
+/**
+ * @brief Given the 3D BB, calculate the wire-length estimate of the net
+ * @param net_id
+ * @param bb
+ * @return
+ */
+static double get_net_wirelength_estimate(ClusterNetId net_id, const t_bb& bb);
+
+/**
+ * @brief Given the per-layer BB, calculate the wire-length estimate of the net on each layer
+ * and return the sum of the lengths
+ * @param net_id
+ * @param bb
+ * @return
+ */
+static double get_net_layer_wirelength_estimate(ClusterNetId /* net_id */,
+                                                const std::vector<t_2D_bb>& bb,
+                                                const vtr::NdMatrixProxy<int, 1> layer_pin_sink_count);
+
+/**
+ * @brief To mitigate round-off errors, every once in a while, the costs of nets are being added from scrath.
+ * This functions is called to do that for bb cost. It doesn't calculate the BBs from scratch, it would only add the costs again.
+ * @return
+ */
+static double recompute_bb_cost();
+
+/**
+ * @brief To get the wirelength cost/est, BB perimiter is multiplied by a factor. This function returns that factor which is a function of net's fan-out.
+ * @return double
+ */
+static double wirelength_crossing_count(size_t fanout);
+
+/**
+ * @breif Calculate the wire-length cost of nets affected by moving the blocks and set bb_delta_c to the total cost change.
```

Comment:
I agree that it would be cleaner to use .size(). If you agree, I'll create an issue regarding these data structures so we can study them later to see the effect of removing the counter on the runtime.

Reply:
.size(): OK to ignore that comment as Amin says it slows things down.